src/cnav_msg.c

```diff
@@ -0,0 +1,515 @@
+  v27_chainback_likely(&part->dec, tmp_bits,
+                       GPS_L2C_V27_DECODE_BITS + GPS_L2C_V27_DELAY_BITS);
+
+  /* Read decoded bits and add them to the decoded buffer */
+  _cnav_bitcopy(part->decoded, part->n_decoded,
+                tmp_bits, 0, GPS_L2C_V27_DECODE_BITS);
+  part->n_decoded += GPS_L2C_V27_DECODE_BITS;
+
+  /* Depending on the decoder state, one of the following actions are
+   * possible:
+   * - If no message lock
+   *   - If no preamble seen - look for preamble
+   *   - If preamble seen - collect 300 bits
+   *     - If 300 bits are collected - verify CRC
+   *       - If CRC is OK - message lock is acquired
+   *       - If CRC fails - rescan for preamble
+   *         - If found - continue collecting 300 bits
+   *         - If not found - continue preamble wait
+   * - If message lock
+   *   - If 300 bits collected, compute CRC
+   *     - If CRC is OK, message can be decoded
+   *     - If CRC is not OK, discard data
+   */
+
+  bool retry = true;
+  while (retry) {
+    retry = false;
+
+    if (!part->preamble_seen) {
+      /* Rescan for preamble if possible. The first bit is ignored. */
+      _cnav_rescan_preamble(part);
+    }
+    if (part->preamble_seen && GPS_CNAV_MSG_LENGTH <= part->n_decoded) {
+
+      /* We have collected 300 bits starting from message preamble. Now try
+       * to compute CRC-24Q */
+      u32 crc  = _cnav_compute_crc(part);
+      u32 crc2 = _cnav_extract_crc(part);
+
+      if (part->message_lock) {
+        /* We have message lock */
+        part->crc_ok = (crc == crc2);
+        if (part->crc_ok) {
+          /* Reset message lock counter */
+          part->n_crc_fail = 0;
+        } else {
+          /* Increment message lock counter */
+          part->n_crc_fail++;
+          if (part->n_crc_fail > GPS_CNAV_LOCK_MAX_CRC_FAILS) {
+            /* CRC has failed too many times - drop the lock. */
+            part->n_crc_fail    = 0;
+            part->message_lock  = false;
+            part->preamble_seen = false;
+            /* Try to find a new preamble, reuse data from buffer. */
+            retry = true;
+          }
+        }
+      } else if (crc == crc2) {
+        /* CRC match - message can be decoded */
+        part->message_lock = true;
+        part->crc_ok       = true;
+        part->n_crc_fail   = 0;
+      } else {
+        /* There is no message lock and the CRC check fails. Assume there is
+         * false positive lock - rescan for preamble. */
+        part->crc_ok = false;
+        part->preamble_seen = false;
+
+        /* CRC mismatch - try to re-scan for preamble */
+        retry = true;
+      }
+    }
+    else
+    {
+      /* No preamble or preamble and less than 300 bits decoded */
+    }
+  }
+}
+
+/**
+ * Invert message bits in the buffer.
+ *
+ * The method inverts bits of the decoded data.
+ *
+ * \param[in,out] part Decoder component with a message buffer.
+ *
+ * \return None
+ */
+void _cnav_msg_invert(cnav_v27_part_t *part)
```

Comment:
Also a candidate to be turned into a generic invert buffer function in `bits.c`?